src/core/TESTING.md

```diff
@@ -59,11 +59,294 @@ test('my test', async () => {
+      something: request.body.text,
+    });
+    return response.ok({ body: { id } });
+  }
+);
+
+router.get(
+  {
+    path: '/myPlugin/something/{id}',
+    validate: {
+      params: schema.object({
+        id: schema.string(),
+      }),
+    },
+  },
+  async (context, request, response) => {
+    try {
+      const savedObjectsClient = context.core.savedObjects.client;
+      const { attributes } = await savedObjectsClient.get('myPlugin-type', request.params.id);
+
+      return response.ok({
+        body: attributes.something,
+      });
+    } catch(error) {
+      if (SavedObjectsErrorHelpers.isNotFoundError(error)) {
+        return response.notFound()
+      }
+      throw e;
+    }
+  }
+);
+```
+You can reuse existing [api_integration](/test/api_integration/config.js) setup by registering a test file within a [test loader](/test/api_integration/apis/index.js). More about the existing FTR setup in the [contribution guide](/CONTRIBUTING.md#running-specific-kibana-tests)
+
+The tests cover:
+- authenticated / non-authenticated user access (when applicable)
+```ts
+// TODO after https://github.com/elastic/kibana/pull/53208/
+```
+- request validation
+```ts
+// test/api_integration/apis/my_plugin/something.ts
+export default function({ getService }: FtrProviderContext) {
+  const supertest = getService('supertest');
+  describe('myPlugin', () => {
+    it('validate params before to store something', async () => {
+      const response = await supertest
+        .post('/myPlugin/something')
+        .set('content-type', 'application/json')
+        .send({ text: 'aaa'.repeat(100) })
+        .expect(400);
+
+      expect(response.body).to.have.property('message');
+      expect(response.body.message).to.contain('must have a maximum length of [100]');
+    });
+  });
+```
+- the main logic of the plugin
+```ts
+export default function({ getService }: FtrProviderContext) {
+  const supertest = getService('supertest');
+  describe('myPlugin', () => {
+    it('stores something', async () => {
+      const response = await supertest
+        .post('/myPlugin/something')
+        .set('content-type', 'application/json')
+        .send({ text: 'aaa' })
+        .expect(200);
+
+      expect(response.body).to.have.property('id');
+      expect(response.body.id).to.be.a('string');
+    });
+
+    it('retrieves something', async () => {
+      const { body } = await supertest
+        .post('/myPlugin/something')
+        .set('content-type', 'application/json')
+        .send({ text: 'bbb' })
+        .expect(200);
+
+      const response = await supertest.get(`/myPlugin/something/${body.id}`).expect(200);
+      expect(response.text).be('bbb');
+    });
+
+    it('returns NotFound error when cannot find a resource', async () => {
+      await supertest
+        .get('/myPlugin/something/missing')
+        .expect(404, 'Saved object [myPlugin-type/missing] not found');
+    });
+  });
+```
 
-#### HTTP Routes
+##### TestUtils
```

Comment:
Although it's a form of integration testing, I wonder if we shouldn't move this out of the "Integration tests" section into it's own section called "Controller testing" or "controller integration testing". ITO the testing pyramid we would have from bottom to top: unit testing, controller testing, integration testing. Because it's a very lightweight type of integration test we should encourage more "controller testing" than fullstack integration tests.

Reply:
`TestUtils` is not limited by `controller testing`. For example, you can use it to test any data kept in memory by your plugin (registry pattern).